homeassistant/components/humidifier/__init__.py

```diff
@@ -0,0 +1,176 @@
+        "async_set_mode",
+        [SUPPORT_MODES],
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_HUMIDITY,
+        make_entity_service_schema({vol.Required(ATTR_HUMIDITY): vol.Coerce(int)}),
+        "async_set_humidity",
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    return await hass.data[DOMAIN].async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    return await hass.data[DOMAIN].async_unload_entry(entry)
+
+
+class HumidifierEntity(ToggleEntity):
+    """Representation of a humidifier device."""
+
+    @property
+    def capability_attributes(self) -> Dict[str, Any]:
+        """Return capability attributes."""
+        supported_features = self.supported_features
+        data = {
+            ATTR_MIN_HUMIDITY: self.min_humidity,
+            ATTR_MAX_HUMIDITY: self.max_humidity,
+        }
+
+        if supported_features & SUPPORT_MODES:
+            data[ATTR_AVAILABLE_MODES] = self.available_modes
+
+        return data
+
+    @property
+    def state_attributes(self) -> Dict[str, Any]:
+        """Return the optional state attributes."""
+        supported_features = self.supported_features
+        data = {
+            ATTR_HUMIDITY: self.target_humidity,
+        }
+
+        if supported_features & SUPPORT_MODES:
+            data[ATTR_MODE] = self.mode
+
+        return data
+
+    @property
+    def target_humidity(self) -> Optional[int]:
+        """Return the humidity we try to reach."""
+        return None
+
+    @property
+    def mode(self) -> Optional[str]:
+        """Return the current mode, e.g., home, auto, baby.
+
+        Requires SUPPORT_MODES.
+        """
+        raise NotImplementedError
+
+    @property
+    def available_modes(self) -> Optional[List[str]]:
+        """Return a list of available modes.
+
+        Requires SUPPORT_MODES.
+        """
+        raise NotImplementedError
+
+    def set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+        raise NotImplementedError()
+
+    async def async_set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+        await self.hass.async_add_executor_job(self.set_humidity, humidity)
+
+    def set_mode(self, mode: str) -> None:
+        """Set new mode."""
+        raise NotImplementedError()
+
+    async def async_set_mode(self, mode: str) -> None:
+        """Set new mode."""
+        await self.hass.async_add_executor_job(self.set_mode, mode)
+
+    @property
+    @abstractmethod
+    def supported_features(self) -> int:
+        """Return the list of supported features."""
+
+    @property
+    def device_class(self) -> str:
+        """Return the device class of the humidifier."""
+        return DEVICE_CLASS_HUMIDIFIER
```

Comment:
I'm actually not entirely sure whether I need to provide this in the base class or not.

Reply:
There's a default `device_class` property in the base `Entity` property that returns `None`. So we don't need it here.

Reply:
Well, I think this default wouldn't hurt. Thanks!

Reply:
We should not have a default device class, please remove.

Reply:
Everything that doesn't have a device class is a humidifier. That's how every other integration works too.